cache/network/api.go

```diff
@@ -0,0 +1,230 @@
+
+	req, err := retryablehttp.NewRequest(uploadMethod, uploadURL, file)
+	if err != nil {
+		return err
+	}
+	for k, v := range headers {
+		req.Header.Set(k, v)
+	}
+
+	resp, err := c.httpClient.Do(req)
+	if err != nil {
+		return err
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		errorResp, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return err
+		}
+		return fmt.Errorf("HTTP %d: %s", resp.StatusCode, errorResp)
+	}
+	return nil
+}
+
+func (c apiClient) acknowledgeUpload(uploadID string) error {
+	url := fmt.Sprintf("%s/upload/%s/acknowledge", c.baseURL, uploadID)
+
+	req, err := retryablehttp.NewRequest("PATCH", url, nil)
+	if err != nil {
+		return err
+	}
+	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", c.accessToken))
+
+	resp, err := c.httpClient.Do(req)
+	if err != nil {
+		return err
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		errorResp, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return err
+		}
+		return fmt.Errorf("HTTP %d: %s", resp.StatusCode, errorResp)
+	}
+	return nil
+}
+
+func (c apiClient) restore(cacheKeys []string) (string, error) {
+	keysInQuery, err := validateKeys(cacheKeys)
+	if err != nil {
+		return "", err
+	}
+	apiURL := fmt.Sprintf("%s/restore?cache_keys=%s", c.baseURL, keysInQuery)
+
+	req, err := retryablehttp.NewRequest("GET", apiURL, nil)
+	if err != nil {
+		return "", err
+	}
+	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", c.accessToken))
+
+	resp, err := c.httpClient.Do(req)
+	if err != nil {
+		return "", err
+	}
+	defer func(Body io.ReadCloser) {
+		err := Body.Close()
+		if err != nil {
+			panic(err)
+		}
+	}(resp.Body)
+
+	if resp.StatusCode == http.StatusNotFound {
+		return "", ErrCacheNotFound
+	}
+	if resp.StatusCode != http.StatusOK {
+		errorResp, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return "", err
+		}
+		return "", fmt.Errorf("HTTP %d: %s", resp.StatusCode, errorResp)
+	}
+
+	var response restoreResponse
+	err = json.NewDecoder(resp.Body).Decode(&response)
+	if err != nil {
+		return "", err
+	}
+
+	return response.URL, nil
+}
+
+func (c apiClient) downloadArchive(url string) (io.ReadCloser, error) {
+	resp, err := c.httpClient.Get(url)
+	if err != nil {
+		return nil, err
+	}
+	if resp.StatusCode != http.StatusOK {
+		defer func(Body io.ReadCloser) {
+			err := Body.Close()
+			if err != nil {
+				panic(err)
+			}
+		}(resp.Body)
+		errorResp, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return nil, err
+		}
+		return nil, fmt.Errorf("HTTP %d: %s", resp.StatusCode, errorResp)
+	}
+
+	return resp.Body, nil
+}
+
+func validateKeys(keys []string) (string, error) {
+	if len(keys) > maxKeyCount {
+		return "", fmt.Errorf("maximum number of keys is %d", maxKeyCount)
+	}
+	var truncatedKeys []string
+	for _, key := range keys {
+		if strings.Contains(key, ",") {
+			return "", fmt.Errorf("commas are not allowed in keys")
+		}
+		if len(key) > maxKeyLength {
+			truncatedKeys = append(truncatedKeys, key[:maxKeyLength])
```

Comment:
Shouldn't it be an error?

Reply:
Since keys are prefix-matched by the backend anyway, it's possible that the truncated key matches a stored cache entry. And overall we try to minimize the hard errors in the cache steps because cache is not a required step in the workflows, just an optimization.